staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go

```diff
@@ -107,6 +107,16 @@ func (w *watcher) Watch(ctx context.Context, key string, rev int64, opts storage
 	if opts.ProgressNotify && w.newFunc == nil {
 		return nil, apierrors.NewInternalError(errors.New("progressNotify for watch is unsupported by the etcd storage because no newFunc was provided"))
 	}
+	// we need the below check because the cacher (ListWatcher) bypasses the REST layer,
+	// so the options are not validated. Without this, we might end up in a situation
+	// where streaming is requested, but the FeatureGate is disabled,
+	// and the bookmark will not be sent
+	//
+	// in such a case, client-go is going to fall back to a standard LIST on any error
+	// returned for watch-list requests
+	if isWatchListRequest(opts) && !utilfeature.DefaultFeatureGate.Enabled(features.WatchList) {
+		return nil, fmt.Errorf("sendInitialEvents is forbidden for watch unless the WatchList feature gate is enabled")
```

Comment:
I'm not sure if this is the best place to add this check. The reasoning in the comment.

Reply:
done.